File: android/app/src/main/java/sol/vortexx/android/files/impl/ResumableUpload.kt
package sol.vortexx.android.files.impl

import io.ktor.client.call.body
import io.ktor.client.request.post
import io.ktor.client.request.put
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.Serializable
import sol.vortexx.android.crypto.api.Aead
import sol.vortexx.android.crypto.util.Hex
import sol.vortexx.android.files.api.TransferProgress
import sol.vortexx.android.keys.api.KeyAcquisition
import sol.vortexx.android.keys.api.RoomKeyProvider
import sol.vortexx.android.net.impl.VortexHttpClient
import java.io.InputStream
import java.security.MessageDigest
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Chunked resumable upload for large files (> 50 MB). Each chunk is
 * independently AES-GCM encrypted so a partial upload doesn't leak the
 * rest of the file if the server aborts mid-stream.
 *
 * Server protocol (mirrors Vortex /api/files/resumable/*):
 *   POST  .../start   → { upload_id }
 *   PUT   .../{id}/chunk?index=N (ciphertext_hex body)
 *   POST  .../{id}/finish → { file_id }
 */
@Singleton
class ResumableUpload @Inject constructor(
    private val http: VortexHttpClient,
    private val aead: Aead,
    private val keys: RoomKeyProvider,
) {
    suspend fun upload(
        roomId: Long,
        source: InputStream,
        filename: String,
        sizeBytes: Long,
        chunkBytes: Int = DEFAULT_CHUNK,
    ): Flow<TransferProgress> = flow {
        val keyAcq = keys.keyFor(roomId)
        if (keyAcq !is KeyAcquisition.Ready) {
            emit(TransferProgress.Error("no_room_key")); return@flow
        }
        val key = Hex.decode(keyAcq.keyHex)

        val startResp = http.client.post("api/files/resumable/start") {
            contentType(ContentType.Application.Json)
            setBody(StartReq(room_id = roomId, filename = filename, size_bytes = sizeBytes))
        }
        if (!startResp.status.isSuccess()) {
            emit(TransferProgress.Error("start_http_${startResp.status.value}")); return@flow
        }
        val uploadId = startResp.body<StartResp>().upload_id

        val md = MessageDigest.getInstance("SHA-256")
        var index = 0
        var done = 0L
        val buf = ByteArray(chunkBytes)

        while (true) {
            val read = source.read(buf)
            if (read <= 0) break
            val chunkPlain = if (read == buf.size) buf else buf.copyOf(read)
            val ct = aead.encrypt(key, chunkPlain)
            md.update(ct)
            val resp = http.client.put("api/files/resumable/$uploadId/chunk?index=$index") {
                contentType(ContentType.Application.OctetStream)
                setBody(Hex.encode(ct))
            }
            if (!resp.status.isSuccess()) {
                emit(TransferProgress.Error("chunk_${index}_http_${resp.status.value}"))
                return@flow
            }
            index += 1
            done  += read
            emit(TransferProgress.InFlight(done = done, total = sizeBytes))
        }

        val finish = http.client.post("api/files/resumable/$uploadId/finish")
        if (!finish.status.isSuccess()) {
            emit(TransferProgress.Error("finish_http_${finish.status.value}")); return@flow
        }
        val body = finish.body<FinishResp>()
        emit(TransferProgress.Done(
            fileId = body.file_id,
            ciphertextHashHex = Hex.encode(md.digest()),
        ))
    }

    companion object { const val DEFAULT_CHUNK = 4 * 1024 * 1024 /* 4 MiB */ }

    @Serializable private data class StartReq(
        val room_id: Long, val filename: String, val size_bytes: Long,
    )
    @Serializable private data class StartResp(val upload_id: String)
    @Serializable private data class FinishResp(val file_id: String)
}

File: ios/Modules/Sources/Files/impl/ResumableUpload.swift
import Foundation
import CryptoKit
import Net
import VortexCrypto
import Keys

/// Chunked resumable upload for large files (> 10 MiB).
///
/// Each chunk is independently AES-GCM encrypted — partial uploads can
/// never leak the rest of the file if the server aborts mid-stream.
/// Protocol matches `scripts/integrity_repo.py` clients in the web flow:
///
///   POST  /api/files/resumable/start            → {upload_id}
///   PUT   /api/files/resumable/{id}/chunk?index → body = ciphertext hex
///   POST  /api/files/resumable/{id}/finish      → {file_id}
public final class ResumableUpload {
    private let http: HttpClient
    private let aead: Aead
    private let keys: RoomKeyProvider
    private let chunkSize: Int

    public init(http: HttpClient, aead: Aead, keys: RoomKeyProvider, chunkSize: Int = 4 * 1024 * 1024) {
        self.http = http
        self.aead = aead
        self.keys = keys
        self.chunkSize = chunkSize
    }

    public func upload(roomId: Int64, filename: String, data: Data) -> AsyncStream<TransferProgress> {
        AsyncStream { cont in
            Task {
                let keyHex: String
                switch await keys.keyFor(roomId) {
                case .ready(let hex, _): keyHex = hex
                case .pending(let r):    cont.yield(.error("no_room_key:\(r)")); cont.finish(); return
                case .error(let r):      cont.yield(.error(r));                  cont.finish(); return
                }
                do {
                    let key = try Hex.decode(keyHex)

                    // 1. Begin session.
                    let start = try await http.send(
                        try HttpRequest.postJson(
                            "api/files/resumable/start",
                            body: StartReq(
                                room_id: roomId, filename: filename,
                                size_bytes: Int64(data.count),
                            ),
                        ),
                        StartResp.self,
                    )

                    // 2. Push chunks.
                    var hasher = SHA256()
                    var index = 0
                    var offset = 0
                    while offset < data.count {
                        let end = min(offset + chunkSize, data.count)
                        let chunk = data.subdata(in: offset..<end)
                        let ct = try aead.encrypt(key: key, plaintext: chunk)
                        hasher.update(data: ct)

                        let hex = Hex.encode(ct)
                        var req = HttpRequest(
                            method: .PUT,
                            path: "api/files/resumable/\(start.upload_id)/chunk",
                            body: hex.data(using: .utf8),
                            extraHeaders: ["Content-Type": "application/octet-stream"],
                            queryItems: [URLQueryItem(name: "index", value: String(index))],
                        )
                        _ = try await http.send(req)

                        offset = end
                        index += 1
                        cont.yield(.inFlight(done: Int64(offset), total: Int64(data.count)))
                    }

                    // 3. Finish session.
                    let finish = try await http.send(
                        HttpRequest(method: .POST,
                                    path: "api/files/resumable/\(start.upload_id)/finish"),
                        FinishResp.self,
                    )
                    cont.yield(.done(
                        fileId: finish.file_id,
                        ciphertextHashHex: Hex.encode(Data(hasher.finalize())),
                        plaintext: nil,
                    ))
                } catch {
                    cont.yield(.error((error as NSError).localizedDescription))
                }
                cont.finish()
            }
        }
    }

    // MARK: DTOs -------------------------------------------------------

    private struct StartReq: Encodable {
        let room_id: Int64
        let filename: String
        let size_bytes: Int64
    }
    private struct StartResp: Decodable  { let upload_id: String }
    private struct FinishResp: Decodable { let file_id: String }
}
